functions/post_data.go

```diff
@@ -0,0 +1,147 @@
+package functions
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+)
+
+const timeoutSeconds = 10 // the max number of seconds to wait for a response
+var errURLMissing = errors.New("error calling PostData; URL is required")
+var errURLEmpty = errors.New("error calling PostData; URL is empty")
+
+// PostData sends an HTTP POST request to a given URL with the data provided,
+// returning either the whole response body or a specified JSON element.
+func PostData(existingHeaders map[string]string, args []string) (string, error) {
+	// Ensure that we have at least one argument for the URL
+	if len(args) == 0 {
+		return "", errURLMissing
+	}
+
+	// Get the URL, response element, and request body from the args
+	url, responseElement, err := argsToURLResponseElement(args)
+	if err != nil {
+		return "", err
+	}
+	var requestBody string
+	if len(args) > 2 {
+		requestBody = argsToRequestBody(existingHeaders, args[2:])
+	}
+
+	// Send the request
+	request, err := http.NewRequest("POST", url, bytes.NewBuffer([]byte(requestBody)))
+	request.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	client := &http.Client{}
+	client.Timeout = time.Second * timeoutSeconds
+	response, err := client.Do(request)
+	if err != nil {
+		return "", err
+	}
+	responseBody, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return "", err
+	}
+	defer response.Body.Close()
+
+	// Get the specific element requested from the response
+	var result string
+	jsonBody := make(map[string]json.RawMessage)
+	err = json.Unmarshal(responseBody, &jsonBody)
+	if err != nil {
+		// This must be an array, so just return the whole response since we can't determine which element to return
+		responseElement = ""
+	}
+	if responseElement != "" {
+		// Return the value associated with the key responseElement,
+		// either a plain string for primitives, or a JSON string for objects/arrays
+		value := jsonBody[responseElement]
+		if value == nil {
+			return "", fmt.Errorf("error: element %s not found in response body", responseElement)
+		}
+		str, err := jsonToString(value)
+		if err != nil {
+			return "", err
+		}
+		result = str
+	} else {
+		// No specific element was requested, so return the whole response body as a JSON string
+		str, err := jsonToString(responseBody)
+		if err != nil {
+			return "", err
+		}
+		result = str
+	}
+
+	return result, nil
+}
+
+// Gets the URL and (optional) responseElement from the args.
+func argsToURLResponseElement(args []string) (string, string, error) {
```

Comment:
Perhaps we  could use a `struct` or names return values for the two strings that are returned. Maybe it's just me, but the comments are the last place I look.:)

Reply:
Thanks. I went with named return values since it's just two strings. I think that does make it clearer.